File: src/ios/AUSTASuperApp/Core/Network/APIEndpoints.swift
//
// APIEndpoints.swift
// AUSTA SuperApp
//
// Created by AUSTA Development Team
// Copyright © 2023 AUSTA Health. All rights reserved.
//

import Foundation // Version: iOS 14.0+

/// Comprehensive API endpoint definitions for the AUSTA SuperApp
/// Implements REST and WebSocket protocols with full security compliance
public enum APIEndpoints {
    
    // MARK: - Authentication Endpoints
    public enum auth {
        /// OAuth2 login with support for multiple authentication methods
        case login(method: String)
        /// New user registration with validation
        case register
        /// JWT token refresh
        case refreshToken
        /// Biometric authentication verification
        case biometricAuth
        /// Multi-factor authentication verification
        case mfaVerify(method: String)
        /// Session management operations
        case sessionManagement(action: String)
        
        var path: String {
            let base = AppConstants.API.ENDPOINTS.AUTH
            switch self {
            case .login(let method):
                return "\(base)/login/\(method)"
            case .register:
                return "\(base)/register"
            case .refreshToken:
                return "\(base)/token/refresh"
            case .biometricAuth:
                return "\(base)/biometric"
            case .mfaVerify(let method):
                return "\(base)/mfa/verify/\(method)"
            case .sessionManagement(let action):
                return "\(base)/session/\(action)"
            }
        }
    }
    
    // MARK: - Virtual Care Endpoints
    public enum virtualCare {
        /// Create new telemedicine session
        case createSession(providerId: String)
        /// Join existing telemedicine session
        case joinSession(sessionId: String)
        /// End active telemedicine session
        case endSession(sessionId: String)
        /// WebSocket connection for real-time communication
        case webSocketConnect(sessionId: String)
        /// Secure file sharing during session
        case fileShare(sessionId: String)
        /// Real-time status updates
        case realTimeStatus(sessionId: String)
        
        var path: String {
            let base = AppConstants.API.ENDPOINTS.VIRTUAL_CARE
            switch self {
            case .createSession(let providerId):
                return "\(base)/sessions/create/\(providerId)"
            case .joinSession(let sessionId):
                return "\(base)/sessions/join/\(sessionId)"
            case .endSession(let sessionId):
                return "\(base)/sessions/end/\(sessionId)"
            case .webSocketConnect(let sessionId):
                return "\(AppConstants.API.WEBSOCKET_URL)/sessions/\(sessionId)"
            case .fileShare(let sessionId):
                return "\(base)/sessions/\(sessionId)/files"
            case .realTimeStatus(let sessionId):
                return "\(base)/sessions/\(sessionId)/status"
            }
        }
    }
    
    // MARK: - Health Records Endpoints
    public enum healthRecords {
        /// Retrieve health records with HIPAA compliance
        case getRecords(patientId: String)
        /// Secure document upload
        case uploadDocument(type: String)
        /// Share records with authorized providers
        case shareRecord(recordId: String, recipientId: String)
        /// Wearable device data synchronization
        case syncWearableData(deviceId: String)
        /// Access audit trail
        case auditTrail(recordId: String)
        /// Encrypted data transfer
        case encryptedTransfer(recordId: String)
        
        var path: String {
            let base = AppConstants.API.ENDPOINTS.HEALTH_RECORDS
            switch self {
            case .getRecords(let patientId):
                return "\(base)/\(patientId)"
            case .uploadDocument(let type):
                return "\(base)/documents/upload/\(type)"
            case .shareRecord(let recordId, let recipientId):
                return "\(base)/\(recordId)/share/\(recipientId)"
            case .syncWearableData(let deviceId):
                return "\(base)/wearables/sync/\(deviceId)"
            case .auditTrail(let recordId):
                return "\(base)/\(recordId)/audit"
            case .encryptedTransfer(let recordId):
                return "\(base)/\(recordId)/transfer"
            }
        }
    }
    
    // MARK: - Insurance Claims Endpoints
    public enum claims {
        /// Submit new insurance claim
        case submitClaim
        /// Retrieve claims history
        case getClaims(status: String?)
        /// Check claim status
        case getClaimStatus(claimId: String)
        /// Upload claim supporting documents
        case uploadClaimDocument(claimId: String)
        /// Batch claims processing
        case batchProcess(batchId: String)
        /// Document verification
        case verifyDocuments(claimId: String)
        
        var path: String {
            let base = AppConstants.API.ENDPOINTS.CLAIMS
            switch self {
            case .submitClaim:
                return "\(base)/submit"
            case .getClaims(let status):
                return status != nil ? "\(base)?status=\(status!)" : base
            case .getClaimStatus(let claimId):
                return "\(base)/\(claimId)/status"
            case .uploadClaimDocument(let claimId):
                return "\(base)/\(claimId)/documents"
            case .batchProcess(let batchId):
                return "\(base)/batch/\(batchId)"
            case .verifyDocuments(let claimId):
                return "\(base)/\(claimId)/verify"
            }
        }
    }
    
    // MARK: - Marketplace Endpoints
    public enum marketplace {
        /// Get available products/services
        case getProducts(category: String?)
        /// Get detailed product information
        case getProductDetails(productId: String)
        /// Process product purchase
        case purchaseProduct(productId: String)
        /// Retrieve order history
        case getOrders(status: String?)
        /// Check product inventory
        case checkInventory(productId: String)
        /// Process payment
        case processPayment(orderId: String)
        
        var path: String {
            let base = AppConstants.API.ENDPOINTS.MARKETPLACE
            switch self {
            case .getProducts(let category):
                return category != nil ? "\(base)/products?category=\(category!)" : "\(base)/products"
            case .getProductDetails(let productId):
                return "\(base)/products/\(productId)"
            case .purchaseProduct(let productId):
                return "\(base)/products/\(productId)/purchase"
            case .getOrders(let status):
                return status != nil ? "\(base)/orders?status=\(status!)" : "\(base)/orders"
            case .checkInventory(let productId):
                return "\(base)/products/\(productId)/inventory"
            case .processPayment(let orderId):
                return "\(base)/orders/\(orderId)/payment"
            }
        }
    }
    
    // MARK: - Helper Methods
    
    /// Constructs full URL for REST endpoints
    public static func getFullURL(for endpoint: String) -> URL? {
        let baseURL = AppConstants.API.BASE_URL
        let version = AppConstants.API.API_VERSION
        let urlString = "\(baseURL)/\(version)\(endpoint)"
        return URL(string: urlString)
    }
    
    /// Constructs WebSocket URL for real-time endpoints
    public static func getWebSocketURL(for endpoint: String) -> URL? {
        let wsURL = AppConstants.API.WEBSOCKET_URL
        let version = AppConstants.API.API_VERSION
        let urlString = "\(wsURL)/\(version)\(endpoint)"
        return URL(string: urlString)
    }
}

File: src/android/app/src/main/kotlin/com/austa/superapp/core/network/ApiEndpoints.kt
package com.austa.superapp.core.network

import com.austa.superapp.core.constants.AppConstants.NETWORK

/**
 * Defines all API endpoint constants for the AUSTA SuperApp Android client.
 * Provides type-safe access to REST and WebSocket endpoints with versioning support.
 * Implements FHIR-compliant healthcare data exchange endpoints.
 */
object ApiEndpoints {

    /**
     * Authentication and authorization endpoints
     */
    object AUTH {
        const val BASE = "/api/${NETWORK.API_VERSION}/auth"
        const val LOGIN = "$BASE/login"
        const val REGISTER = "$BASE/register"
        const val REFRESH_TOKEN = "$BASE/refresh"
        const val FORGOT_PASSWORD = "$BASE/forgot-password"
        const val RESET_PASSWORD = "$BASE/reset-password"
        const val VERIFY_EMAIL = "$BASE/verify-email"
        const val BIOMETRIC_AUTH = "$BASE/biometric"
        const val LOGOUT = "$BASE/logout"
        const val MFA = "$BASE/mfa"
        const val MFA_VERIFY = "$BASE/mfa/verify"
        const val SESSION_INFO = "$BASE/session"
        const val REVOKE_TOKEN = "$BASE/revoke"
    }

    /**
     * Virtual care and telemedicine endpoints including WebSocket support
     */
    object VIRTUAL_CARE {
        const val BASE = "/api/${NETWORK.API_VERSION}/virtual-care"
        const val SESSIONS = "$BASE/sessions"
        const val JOIN = "$BASE/sessions/{sessionId}/join"
        const val LEAVE = "$BASE/sessions/{sessionId}/leave"
        const val CHAT = "$BASE/sessions/{sessionId}/chat"
        const val PRESCRIPTIONS = "$BASE/sessions/{sessionId}/prescriptions"
        const val NOTES = "$BASE/sessions/{sessionId}/notes"
        const val WS_ENDPOINT = "/ws/sessions/{sessionId}"
        const val MEDIA_CONTROL = "$BASE/sessions/{sessionId}/media"
        const val RECORDING = "$BASE/sessions/{sessionId}/recording"
        const val PARTICIPANTS = "$BASE/sessions/{sessionId}/participants"
        const val QUALITY_METRICS = "$BASE/sessions/{sessionId}/metrics"
        const val SCREEN_SHARE = "$BASE/sessions/{sessionId}/screen"
    }

    /**
     * Health records management endpoints with FHIR support
     */
    object HEALTH_RECORDS {
        const val BASE = "/api/${NETWORK.API_VERSION}/health-records"
        const val FHIR = "$BASE/fhir/r4"
        const val LIST = "$BASE/list?page={page}&size={size}&sort={sort}"
        const val UPLOAD = "$BASE/upload"
        const val BULK_UPLOAD = "$BASE/upload/bulk"
        const val DOWNLOAD = "$BASE/records/{recordId}/download"
        const val SHARE = "$BASE/records/{recordId}/share"
        const val DELETE = "$BASE/records/{recordId}"
        const val CATEGORIES = "$BASE/categories"
        const val SEARCH = "$BASE/search?query={query}&filter={filter}"
        const val SYNC = "$BASE/sync"
        const val AUDIT = "$BASE/records/{recordId}/audit"
        const val PERMISSIONS = "$BASE/records/{recordId}/permissions"
        const val VERSIONS = "$BASE/records/{recordId}/versions"
    }

    /**
     * Insurance claims processing endpoints
     */
    object CLAIMS {
        const val BASE = "/api/${NETWORK.API_VERSION}/claims"
        const val SUBMIT = "$BASE/submit"
        const val BULK_SUBMIT = "$BASE/submit/bulk"
        const val LIST = "$BASE/list?status={status}&date={date}"
        const val STATUS = "$BASE/claims/{claimId}/status"
        const val DOCUMENTS = "$BASE/claims/{claimId}/documents"
        const val HISTORY = "$BASE/history?page={page}&size={size}"
        const val COVERAGE = "$BASE/coverage"
        const val VERIFY = "$BASE/claims/{claimId}/verify"
        const val RESUBMIT = "$BASE/claims/{claimId}/resubmit"
        const val APPEAL = "$BASE/claims/{claimId}/appeal"
        const val ANALYTICS = "$BASE/analytics"
    }

    /**
     * Digital marketplace endpoints for healthcare products and services
     */
    object MARKETPLACE {
        const val BASE = "/api/${NETWORK.API_VERSION}/marketplace"
        const val PRODUCTS = "$BASE/products?category={category}&sort={sort}"
        const val CATEGORIES = "$BASE/categories"
        const val SEARCH = "$BASE/search?query={query}&filter={filter}"
        const val ORDERS = "$BASE/orders"
        const val ORDER_STATUS = "$BASE/orders/{orderId}/status"
        const val CART = "$BASE/cart"
        const val WISHLIST = "$BASE/wishlist"
        const val REVIEWS = "$BASE/products/{productId}/reviews"
        const val RECOMMENDATIONS = "$BASE/recommendations"
        const val PROMOTIONS = "$BASE/promotions"
        const val INVENTORY = "$BASE/products/{productId}/inventory"
        const val BULK_ORDER = "$BASE/orders/bulk"
    }

    /**
     * Utility function to replace path parameters in endpoint URLs
     *
     * @param endpoint The endpoint URL containing parameters
     * @param params Map of parameter names to their values
     * @return The endpoint URL with parameters replaced
     */
    fun buildUrl(endpoint: String, params: Map<String, String>): String {
        var result = endpoint
        params.forEach { (key, value) ->
            result = result.replace("{$key}", value)
        }
        return result
    }
}